internal/provider/resource_gitlab_service_emails_on_push.go

```diff
@@ -0,0 +1,167 @@
+			"project": {
+				Description:  "ID of the project you want to activate integration on.",
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"recipients": {
+				Description:  "Emails separated by whitespace.",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"disable_diffs": {
+				Description: "Disable code diffs.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+			},
+			"send_from_committer_email": {
+				Description: "Send from committer.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+			},
+			"push_events": {
+				Description: "Enable notifications for push events.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+			},
+			"tag_push_events": {
+				Description: "Enable notifications for tag push events.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+			},
+			"branches_to_be_notified": {
+				Description: `Branches to send notifications for. Valid options are "all", "default", "protected", and "default_and_protected". Notifications are always fired for tag pushes.`,
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "all",
+			},
+			"title": {
+				Description: "Title of the integration.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"created_at": {
+				Description: "The ISO8601 date/time that this integration was activated at in UTC.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"updated_at": {
+				Description: "The ISO8601 date/time that this integration was last updated at in UTC.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"slug": {
+				Description: "The name of the integration in lowercase, shortened to 63 bytes, and with everything except 0-9 and a-z replaced with -. No leading / trailing -. Use in URLs, host names and domain names.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"active": {
+				Description: "Whether the integration is active.",
+				Type:        schema.TypeBool,
+				Computed:    true,
+			},
+		},
+	}
+})
+
+func resourceGitlabServiceEmailsOnPushCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	d.SetId(project)
+
+	options := &gitlab.SetEmailsOnPushServiceOptions{
+		Recipients:             gitlab.String(d.Get("recipients").(string)),
+		DisableDiffs:           gitlab.Bool(d.Get("disable_diffs").(bool)),
+		SendFromCommitterEmail: gitlab.Bool(d.Get("send_from_committer_email").(bool)),
+		PushEvents:             gitlab.Bool(d.Get("push_events").(bool)),
+		TagPushEvents:          gitlab.Bool(d.Get("tag_push_events").(bool)),
+		BranchesToBeNotified:   gitlab.String(d.Get("branches_to_be_notified").(string)),
```

Comment:
🚧 **Suggestion:** for better compatibility we usually leave the defaults up to the GitLab API. Just setting all the fields in the request may lead to incpompatibilies between the defaults in the provider and the API.

You can use the following pattern to construct the options:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/96e5f299d937aa1d74281aa21cbb92938f286cbe/internal/provider/resource_gitlab_cluster_agent_token.go#L45

Reply:
cool, didn't know about this. Adapted the code accordingly